src/librustc/hir/lowering.rs

```diff
@@ -1009,7 +1010,7 @@ impl<'a> LoweringContext<'a> {
                         hir::TyPath(hir::QPath::Resolved(None, P(hir::Path {
                             span,
                             def: Def::TyParam(DefId::local(def_index)),
-                            segments: vec![].into(),
+                            segments: hir_vec![hir::PathSegment::from_name(name)],
```

Comment:
I'm intrigued about the implications elsewhere in the compiler of having a path with a single path segment of `impl Bound1 + Bound2`, including that it'd be different from `impl Bound2 + Bound1` (but that is already an issue elsewhere, IIRC) and the high likelihood of code not expecting this.

That being said, the code looks otherwise correct. Deferring to @cramertj and @arielb1, @nikomatsakis as they have more context, but it LGTM.

Reply:
@estebank It's also interesting because there can be multiple different types with the same name (since `fn foo(_: impl Into<u8>, _: impl Into<u8>) { ... }` desugars to something like `fn foo<"impl Into<u8>": Into<u8>, "impl Into<u8>": Into<u8>>(...) { ... }`). I *think* it should be fine, though.

Reply:
That is a good point as well. If there isn't already, I can add a compile-pass test for the `fn f(impl Trait, impl Trait)` with two distinct types to the two arguments. That'll make sure it works.